### PR TITLE
Next Gen GitHub Tag Action

### PR DESCRIPTION
**Pull request by beholdenkey**
# Summary of changes

This is a first iteration refactor. A lot of the code has been modified and updated. However, there is still a lot of room for improvement. I have a dev branch where I am testing out switching from the node server to using Python for the GitPython and SemVer Libraries. That will offer a greater level of performance and extensibility.

In addition, there is now a publish.yml that will build and publish the container image to the GitHub container registry. I saw this was something brought up in previous pull requests. I still need to tie the action into using that image, but it should help with the action time consumed. 

## Breaking Changes

There are not necessarily breaking changes. The functionality of the action has stayed the same; however,r a lot of the code supporting it has.

## How changes have been tested

- Local Unit Tests, Testing the container build/workflows; this will be the first iteration of the pull request to the master origin.

## List any unknowns

- Lots of changes/lots of unknowns. This will more than likely require iterations to complete changes and successfully merge.

### CODE REVIEW

File: action.yml [hunk 1]
```diff
@@ -1,16 +1,27 @@
...
+author: "Nick Sjostrom"
+
+inputs:
+  DEFAULT_BUMP:
+    description: "Default semver bump type (major, minor, patch, or none)"
+    required: false
+    default: "minor"
+  DEFAULT_BRANCH:
+    description: "The default branch for the repository"
+    required: false
```
Comment (matthiasgubler) [inline action.yml:12]: Question: why are you adding these two values as inputs when env-vars provide everything else? Keeping this way would mean, to adjust the examples to something like:
```
      - name: Bump version and push tag
        uses: anothrNick/github-tag-action@1.61.0 # Don't use @master unless you're happy to test the latest version
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          WITH_V: true
       with:
          DEFAULT_BUMP: "patch"
```

As the inputs need to be provided in a different matter. Also accessing the values would need to change to:
```
${{ inputs.DEFAULT_BUMP }}
```

Last but not least, they should be renamed to kebab-case `default-bump`.

I would delete lines 5 - 12 in this case

**Review by sbe-arg — COMMENTED** [review]
Thank you for this. On the phone it looks a considerable improvement in practice.


I have 0 time to check this pr this week but maybe @sammcj can have a look and test it.

File: .github/workflows/publish.yml [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  build:
+    runs-on: ubuntu-latest
+    permissions:
+      contents: read
+      packages: write
+      id-token: write
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v3
+      - name: Install cosign
```
Comment (sammcj) [inline .github/workflows/publish.yml:24]: Oh nice! Great to see some security improvements!

File: .github/workflows/scripts/tag_check.sh [hunk 3]
```diff
@@ -0,0 +1,45 @@
+#!/bin/bash
```
Comment (sammcj) [inline .github/workflows/scripts/tag_check.sh:1]: `!/usr/bin/env bash`

Just so it's portable.

File: .github/workflows/scripts/tag_check.sh [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+
+set -euo pipefail
```
Comment (sammcj) [inline .github/workflows/scripts/tag_check.sh:3]: A nice little addition I like to add to shell scripts to add set -x only if you're debugging:

```bash
DEBUG=${DEBUG:-false}

function debug() {
  echo "Debug mode enabled" >>"$GITHUB_STEP_SUMMARY"

  # Print all environment variables
  env | sort >>"$GITHUB_STEP_SUMMARY"
  echo -e "\n" >>"$GITHUB_STEP_SUMMARY"

	set -x
}

if [[ $DEBUG == 'true' ]]; then
  debug
fi
```

File: .github/workflows/scripts/tag_check.sh [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+PRE_OUTPUT_NEWTAG=$5
+PRE_OUTPUT_PART=$6
+
+echo "Outputs from running the action:" >>"$GITHUB_STEP_SUMMARY"
+echo "MAIN Tag: $MAIN_OUTPUT_TAG" >>"$GITHUB_STEP_SUMMARY"
+echo "MAIN New tag: $MAIN_OUTPUT_NEWTAG" >>"$GITHUB_STEP_SUMMARY"
+echo "MAIN Part: $MAIN_OUTPUT_PART" >>"$GITHUB_STEP_SUMMARY"
+echo "PRE Tag: $PRE_OUTPUT_TAG" >>"$GITHUB_STEP_SUMMARY"
+echo "PRE New tag: $PRE_OUTPUT_NEWTAG" >>"$GITHUB_STEP_SUMMARY"
+echo "PRE Part: $PRE_OUTPUT_PART" >>"$GITHUB_STEP_SUMMARY"
```
Comment (sammcj) [inline .github/workflows/scripts/tag_check.sh:18]: A little trick with this to make it render in a nice table can be something like this:

```bash
OUTPUT_SUMMARY="\n
| Output       | Value                 | \n
| ------------ | --------------------- | \n
|              |                       | \n
| MAIN Tag     | ${MAIN_OUTPUT_TAG}    | \n
| MAIN New tag | ${MAIN_OUTPUT_NEWTAG} | \n
| MAIN Part    | ${MAIN_OUTPUT_PART}   | \n
| PRE Tag      | ${PRE_OUTPUT_TAG}     | \n
| PRE New tag  | ${PRE_OUTPUT_NEWTAG}  | \n
| PRE Part     | ${PRE_OUTPUT_PART}    | \n
"

echo -e "$OUTPUT_SUMMARY" >>"$GITHUB_STEP_SUMMARY"

File: CODE_OF_CONDUCT.md [hunk 6]
```diff
@@ -1,72 +1,45 @@
...
-* Focusing on what is best not just for us as individuals, but for the
-  overall community
+- Demonstrating empathy and kindness toward others
+- Being respectful of differing opinions, viewpoints, and experiences
+- Giving and gracefully accepting constructive feedback
+- Accepting responsibility, apologizing to those affected by our mistakes, and learning from the experience
+- Focusing on what is best not only for us as individuals but also for the overall community
 
-Examples of unacceptable behaviour include:
+Examples of unacceptable behavior include:
```
Comment (sammcj) [inline CODE_OF_CONDUCT.md:17]: Behaviour is the correct international spelling.

Behavior is specific to American.

File: .github/workflows/scripts/tag_check.sh [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+if $main && $pre; then
+	echo "The tags were created correctly" >>"$GITHUB_STEP_SUMMARY"
+else
+	echo "Tags not created correctly" >>"$GITHUB_STEP_SUMMARY"
+	exit 1
+fi
+
+# Test for #none bump
+if [[ $MAIN_OUTPUT_PART == "none" ]]; then
+	if [[ $MAIN_OUTPUT_TAG == "$MAIN_OUTPUT_NEWTAG" ]]; then
```
Comment (sammcj) [inline .github/workflows/scripts/tag_check.sh:39]: I think the variable on the left of the comparison should be quoted as well.

File: .github/workflows/scripts/tag_check.sh [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+verlte() {
+	[ "$1" = "$(echo -e "$1\n$2" | sort -V | head -n1)" ]
+}
+verlt() {
+	[ "$1" = "$2" ] && return 1 || verlte "$1" "$2"
+}
+
+main="$(verlt "$MAIN_OUTPUT_TAG" "$MAIN_OUTPUT_NEWTAG" && true || false)"
+pre="$(verlt "$PRE_OUTPUT_TAG" "$PRE_OUTPUT_NEWTAG" && true || false)"
```
Comment (sammcj) [inline .github/workflows/scripts/tag_check.sh:28]: It feels weird to have double quotes inside a double quoted string but it does seem to work on my machine as well 🤷

**Review by sammcj — CHANGES_REQUESTED** [review]
Awesome effort! Really appreciate the contributions.

One thing I'd say especially given this is a big PR - it should really include some basic tests.

It doesn't have to be some big fancy testing framework but at least a script or functions that validate several scenarios that this may be used for.

File: entrypoint.sh [hunk 9]
```diff
@@ -2,262 +2,227 @@
...
-then
-    new="$custom_tag"
+if [ -n "$custom_tag" ]; then
+	new_tag="$custom_tag"
 fi
 
-# set outputs
-setOutput "new_tag" "$new"
-setOutput "part" "$part"
-setOutput "tag" "$new" # this needs to go in v2 is breaking change
```
Comment (sbe-arg) [inline entrypoint.sh:223]: I would like this comment to remain specially since broke the flows for a LOT lots of people

**Review by sbe-arg — CHANGES_REQUESTED** [review]

**Comment by sbe-arg** [comment]
:(

**Pull request closed without merging**